### PR TITLE
GODRIVER-2690 Set minimum supported Go version to 1.18 in Go Driver v2.0

### PR DESCRIPTION
**Pull request by prestonvasquez**
GODRIVER-2690

## Summary

Migrate the Go imports and module to v2 and set minimum supported Go Version to 1.18. Additionally, updating to 1.18 allows us to remove the vendor directory.

## Background & Motivation

The Go Driver team intends for the master branch to work as the "development" branch. Therefore, a "v1" branch has been cut where v1.x-focused tickets will be development on the "v1" branch and then cherry-picked , as needed, to the master branch. 

This ticket updates the master branch to reflect the "development" branch as v2-focused.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,8 +1,8 @@
 <p align="center"><img src="etc/assets/mongo-gopher.png" width="250"></p>
 <p align="center">
-  <a href="https://goreportcard.com/report/go.mongodb.org/mongo-driver"><img src="https://goreportcard.com/badge/go.mongodb.org/mongo-driver"></a>
-  <a href="https://pkg.go.dev/go.mongodb.org/mongo-driver/mongo"><img src="etc/assets/godev-mongo-blue.svg" alt="docs"></a>
-  <a href="https://pkg.go.dev/go.mongodb.org/mongo-driver/bson"><img src="etc/assets/godev-bson-blue.svg" alt="docs"></a>
+  <a href="https://goreportcard.com/report/go.mongodb.org/mongo-driver/v2"><img src="https://goreportcard.com/badge/go.mongodb.org/mongo-driver/v2"></a>
```
Comment (prestonvasquez) [inline README.md:3]: Do we want to update the data in the README.md file with v2 references? Or do we want to keep the v1 references until v2 releases for user ease-of access?

File: benchmark/operation_test.go [hunk 2]
```diff
@@ -41,7 +41,7 @@ func BenchmarkClientWrite(b *testing.B) {
 			if err != nil {
 				b.Fatalf("error connecting: %v", err)
 			}
-			defer client.Disconnect(context.Background())
+			defer func() { _ = client.Disconnect(context.Background()) }()
```
Comment (prestonvasquez) [inline benchmark/operation_test.go:44]: golangci-lint wants us to check errors.

File: benchmark/bson.go [hunk 3]
```diff
@@ -11,7 +11,7 @@ import (
 	"io/ioutil"
 	"path/filepath"
 
-	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/v2/bson"
```
Comment (matthewdale) [inline benchmark/bson.go:14]: I'm concerned that doing this package major version bump now is going to make keeping v1 and master in-sync difficult for the duration of v2 development. That may cause every v1 --> master sync to hit a merge conflict or require manual modification of the import path. Is it possible to defer making the module version change until later?

**Comment by matthewdale** [comment]
That's quite a diff summary
Screenshot 2023-09-15 at 3 22 41 PM

🎉

File: benchmark/bson.go (hunk 3 above)
Comment (prestonvasquez) [reply benchmark/bson.go:14]: Yep, this makes sense. I will revert those changes.

File: Makefile [hunk 4]
```diff
@@ -52,19 +52,15 @@ install-lll:
...
+# check-modules runs "go mod tidy"  and exits with a non-zero exit code if there
+# are any module changes. The intent is to confirm that exactly the required 
+# modules are declared as dependencies. We should always be able to run "go mod 
+# tidy" and expect that no unrelated changes are made to the "go.mod" file.
 .PHONY: check-modules
 check-modules:
 	go mod tidy -v
-	go mod vendor
-	git diff --exit-code go.mod go.sum ./vendor
+	go mod verify
```
Comment (matthewdale) [inline Makefile:62]: Optional: `go mod verify` seems to conflict with using Go workspaces in some cases (see https://github.com/golang/go/issues/62663). When I run that command on my workstation, I get an error:
```
❯ go mod verify
go.mongodb.org/mongo-driver : missing ziphash: open hash: no such file or directory
```
Consider removing `go mod verify`.

**Review by matthewdale — APPROVED** [review]
Looks good 👍

**Review by matthewdale — APPROVED** [review]
Looks good 👍

**Review by blink1073 — APPROVED** [review]

**Pull request merged**